pkg/key/key.go

```diff
@@ -1,22 +1,109 @@
 package key
 
 import (
+	"context"
 	"fmt"
+	"io/ioutil"
+	"os"
 
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	chartv1 "github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	ClusterNameLabel        = "cluster.x-k8s.io/cluster-name"
-	ClusterWatchFilterLabel = "cluster.x-k8s.io/watch-filter"
+	ClusterNameLabel = "cluster.x-k8s.io/cluster-name"
 
 	FinalizerName = "encryption-provider-operator.finalizers.giantswarm.io"
+
+	MasterNodeLabel = "node-role.kubernetes.io/master"
 )
 
-func GetClusterIDFromLabels(t v1.ObjectMeta) string {
+func GetClusterIDFromLabels(t metav1.ObjectMeta) string {
 	return t.GetLabels()[ClusterNameLabel]
 }
 
 func SecretName(clusterName string) string {
 	return fmt.Sprintf("%s-encryption-provider-config", clusterName)
 }
+
+// GetWCK8sClient will return workload cluster k8s controller-runtime client
+func GetWCK8sClient(ctx context.Context, ctrlClient client.Client, clusterName string, clusterNamespace string) (client.Client, error) {
+	var err error
+
+	if _, err := os.Stat(tempKubeconfigFileName(clusterName)); err == nil {
+		// kubeconfig file already exists, no need to fetch and write again
+
+	} else if os.IsNotExist(err) {
+		var kubeconfig []byte
+		// kubeconfig dont exists we need to fetch it and write to file
+		var secret corev1.Secret
+		{
+			err = ctrlClient.Get(ctx, client.ObjectKey{
+				Name:      fmt.Sprintf("%s-kubeconfig", clusterName),
+				Namespace: clusterNamespace,
+			},
+				&secret)
+
+			if apierrors.IsNotFound(err) {
+				// in legacy gs the kubeconfig is in namespace equal to cluster ID
+				err = ctrlClient.Get(ctx, client.ObjectKey{
+					Name:      fmt.Sprintf("%s-kubeconfig", clusterName),
+					Namespace: clusterName,
+				},
+					&secret)
+				if err != nil {
+					return nil, err
```

Comment:
would be much better to mask errors with `microerror.Mask(err)` everywhere for easier troubleshooting of errors

Reply:
AFAIK  we tend not to use microerror in kubebuilder operators but maybe I am wrong, WDYT @fiunchinho ?

Reply:
There is currently no consensus, but we use it [on some](https://github.com/giantswarm/dns-operator-azure/blob/main/controllers/azurecluster_controller.go#L98). I'd use it

Reply:
ack